lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -24,6 +24,149 @@
   [src getCharacters:(unichar *)&result[0] range:NSMakeRange(0, size)];
   return result;
 }
+const std::vector<std::u16string>& getAvailableLocalesVector() {
+    static const std::vector<std::u16string> *availableLocalesVector = []{
+      NSArray<NSString *> *availableLocales = [NSLocale availableLocaleIdentifiers];
+      // Intentionally leaked to avoid destruction order problems.
+      auto *vec = new std::vector<std::u16string>();
+      for (id str in availableLocales)
+        vec->push_back(nsStringToU16String(str));
+      return vec;
+      }();
+    return *availableLocalesVector;
+}
+std::u16string getDefaultLocale() {
+  NSString *defLocale = [[NSLocale currentLocale] localeIdentifier];
+  return nsStringToU16String(defLocale);
+}
+// Implementer note: This method corresponds roughly to
+// https://tc39.es/ecma402/#sec-bestavailablelocale
+std::u16string bestAvailableLocale(
+    const std::vector<std::u16string> &availableLocales,
+    const std::u16string &locale) {
+  // 1. Let candidate be locale
+  std::u16string candidate = locale;
+
+  // 2. Repeat
+  while (true) {
+    // a. If availableLocales contains an element equal to candidate, return
+    // candidate.
+    if (llvh::find(availableLocales, candidate) !=
+        availableLocales.end()) {
+      return candidate;
+    }
+
+    // b. Let pos be the character index of the last occurrence of "-" (U+002D)
+    // within candidate.
+    size_t pos = candidate.rfind(u'-');
+
+    // ...If that character does not occur, return undefined.
+    if (pos == std::string::npos) {
+        return u"und";
+    }
+
+    // c. If pos ≥ 2 and the character "-" occurs at index pos-2 of candidate,
+    // decrease pos by 2.
+    if (pos >= 2 && candidate[pos - 2] == '-') {
+      pos -= 2;
+    }
+    // d. Let candidate be the substring of candidate from position 0,
+    // inclusive, to position pos, exclusive.
+    candidate.resize(pos);
+  }
+}
+
+// Implementer note: For more information review
+// https://402.ecma-international.org/7.0/#sec-unicode-locale-extension-sequences
+std::u16string toNoExtensionsLocale(const std::u16string &locale) {
+  std::vector<std::u16string> subtags;
+  auto s = locale.begin(), e = locale.end();
+  while (true) {
+    auto tagEnd = std::find(s, e, u'-');
+    subtags.emplace_back(s, tagEnd);
+    if (tagEnd == e)
+      break;
+    s = tagEnd + 1;
+  }
+  std::u16string result;
+  size_t size = subtags.size();
+  for (size_t s = 0; s < size;) {
+    result.append(subtags[s]);
+    s++;
```

Comment:
```suggestion
    s++;
    if (s < size) {
      result.append(u"-");
    }
```

Reply:
I think we should move this to the start of the loop, and instead test if `s > 0`. As written, this will not correctly handle the case where there is a trailing unicode extension right? I'm thinking that something like `tr-TR-u-banana` will become `tr-TR-`.